jbi/retry.py

```diff
@@ -0,0 +1,84 @@
+import asyncio
+import logging
+from datetime import datetime, timedelta
+from os import getenv
+from time import sleep
+from typing import AsyncIterator
+
+import jbi.runner as runner
+from jbi.configuration import ACTIONS
+from jbi.queue import QueueItem, get_dl_queue
+
+CONSTANT_RETRY = getenv("DL_QUEUE_CONSTANT_RETRY", "false") == "true"
+RETRY_TIMEOUT_DAYS = getenv("DL_QUEUE_RETRY_TIMEOUT_DAYS", 7)
+
+
+async def retry_failed(item_executor=runner.execute_action):
+    queue = get_dl_queue()
+    logger = logging.getLogger(__name__)
+    min_event_timestamp = datetime.now() - timedelta(days=int(RETRY_TIMEOUT_DAYS))
+
+    # load all bugs from DLQ
+    bugs: dict[int, AsyncIterator[QueueItem]] = await queue.retrieve()
+
+    # metrics to track
+    metrics = {
+        "bug_count": len(bugs),
+        "events_processed": 0,
+        "events_skipped": 0,
+        "events_failed": 0,
+        "bugs_failed": 0,
+    }
+
+    for bug_id, items in bugs.items():
+        try:
+            async for item in items:
+                # skip and delete item if we have exceeded max_timeout
+                if item.timestamp < min_event_timestamp:
+                    logger.warning("removing expired event %s", item.identifier)
+                    await queue.done(item)
+                    metrics["events_skipped"] += 1
+                    continue
+
+                try:
+                    item_executor(item.payload, ACTIONS)
+                    await queue.done(item)
+                    metrics["events_processed"] += 1
+                except Exception as ex:
+                    # write well formed log that could be alerted on
+                    logger.error(
+                        "failed to reprocess event %s. error: %s", item.identifier, ex
+                    )
+                    metrics["events_failed"] += 1
+
+                    # check for other events that will be skipped
+                    skipped_events = await queue.list(bug_id)
```

Comment:
A mocking gotcha. Somehow we didn't catch that `queue.list` isn't a method (an oversight on my part), rather it's `queue.backend.list`. I think creating a MagicMock with [`spec`](https://docs.python.org/3/library/unittest.mock.html#auto-speccing). will help us avoid things like this.

Reply:
[1ee3ab7](https://github.com/mozilla/jira-bugzilla-integration/pull/918/commits/1ee3ab7f921311e094a0b5ce5211730c40097fde)